packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,78 +86,107 @@ export class WorkspaceService implements FrontendApplicationContribution {
         return this.onWorkspaceChangeEmitter.event;
     }
 
+    protected async setWorkspaceFile(newWorkspaceFile: FileStat | undefined): Promise<void> {
+        if (newWorkspaceFile) {
+            if (this._workspaceFile &&
+                this._workspaceFile.uri !== newWorkspaceFile.uri &&
+                this.workspaceFileWatcher) {
+                this.workspaceFileWatcher.dispose();
+                this.workspaceFileWatcher = undefined;
+            }
+            this._workspaceFile = newWorkspaceFile;
+        }
+        this.updateTitle();
+        if (this._workspaceFile) {
+            if (!this.workspaceFileWatcher) {
+                this.workspaceFileWatcher = await this.watcher.watchFileChanges(new URI(this._workspaceFile.uri));
+            }
+            this.watcher.onFilesChanged(event => {
```

Comment:
it should be installed only once in `init`

Reply:
`init()` is not called when the user does a "Save Workspace As...", while we will need to watch the new workspace file in this case. So maybe i should leave `watcher.onFilesChanged()` in this function? This function will only be called in the following two cases:
- theia loads
- "Save workspace As..."

Reply:
Why is it important? We will always watch for file changes and if `this._workspaceFile` is undefined just ignore them. Otherwise, you need to take care of `onFilesChanged` listeners, with the current there are leaking.

Reply:
think about users doing this:
1. opens a workspace that has a workspace file (say, `a.code-workspace`), the watcher is added to that workspace file, everything is fine
2. user clicks "Save workspace As...", and chooses to save the workspace into `b.code-workspace`. now we are still watching `a.code-workspace`, however the workspace has switched from `a` to `b`
3. user adds one more folder to the current workspace, which is `b`. since `b` is not watched, theia does not get the change

that's why i said i needed to call `this.watcher.onFilesChanged()` in `setWorkspaceFile()`

Reply:
`this.watcher.onFilesChanged` reports events from all watchers
you are right you need to start watching of a new workspace file with `watchFileChanges` but you don't need to add another listener to `this.watcher.onFilesChanged`, one is enough